src/scheduler/server_info.c

```diff
@@ -1977,13 +1977,13 @@ update_server_on_end(status *policy, server_info *sinfo, queue_info *qinfo,
 	if (cstat.preempting && resresv->is_job) {
 		if (sinfo->has_soft_limit || resresv->job->queue->has_soft_limit) {
 			for (i = 0; sinfo->jobs[i] != NULL; i++) {
-				if (sinfo->jobs[i]->job !=NULL) {
-					if (((resresv->job->queue->has_user_limit || sinfo->has_user_limit) &&
-					    (!strcmp(resresv->user, sinfo->jobs[i]->user))) ||
-					    ((resresv->job->queue->has_grp_limit || sinfo->has_grp_limit) &&
-					    (!strcmp(resresv->group, sinfo->jobs[i]->group))) ||
-					    ((resresv->job->queue->has_proj_limit || sinfo->has_proj_limit) &&
-					    (!strcmp(resresv->project, sinfo->jobs[i]->project))))
+				if (sinfo->jobs[i]->job != NULL) {
+					int usrlim = resresv->job->queue->has_user_limit | sinfo->has_user_limit;
+					int grplim = resresv->job->queue->has_grp_limit | sinfo->has_grp_limit;
+					int projlim = resresv->job->queue->has_proj_limit | sinfo->has_proj_limit;
```

Comment:
I think we want to use || instead of the bitwise |.  This is because we're using the structure bit fields instead of doing it ourselves.  This only works because has_use_limit is :1.  If it was :2 or :3, it'd hold values that were not bits.

Reply:
You are right Bhroam, but do you see the need for has_user_limit flags to support more than one bit in future? I can not imagine what it would store.
Even if it does, all we are looking for is that if there is any user limit set on queue or server so as long as OR value is a non zero it would work here.

Reply:
structure bitfields are not bits.  They are variables smaller than a byte.  If we were change has_user_limit and turn it into a :4, it just means the value that can be stored can have 4 bits (0-15).  This isn't the same as having a 2 or 4 or 8 stored in an unsigned integer which to use the bitwise operators on.  It just so happens that has_user_limits is a :1, meaning it can contain one bit (0-1).  They aren't meant to be bits, they are meant to be normal variables with smaller ranges.

Reply:
I know that it just limits the range of values a variable can store. But does it matter to us if we are doing 3 | 5 or 3||5 because as long as it is non-zero we know there is a limit set. right?

Reply:
All I am saying is we should use the correct operator.  The '|' operator is meant for bits.  The || operator is meant for normal variables.  A structure bit field is a normal variable with a limited range.  You are correct that both will work, but I don't think the '|' operator is right.  We wouldn't store a set of bits in a bitfield.  We'd use an unsigned long and use the '&' and '|' operators.

Reply:
I've made the change @bhroam and pushed the change.